system/security/securesocket/securesocket.cpp

```diff
@@ -1641,3 +1642,58 @@ SECURESOCKET_API int signCertificate(const char* csr, const char* ca_certificate
 }
 
 }
+
+class CSecureSmartSocketFactory : public CSmartSocketFactory
+{
+public:
+    Owned<ISecureSocketContext> secureContext;
+
+    CSecureSmartSocketFactory(const char *_socklist, bool _retry, unsigned _retryInterval, unsigned _dnsInterval) : CSmartSocketFactory(_socklist, _retry, _retryInterval, _dnsInterval)
+    {
+        secureContext.setown(createSecureSocketContext(ClientSocket));
+    }
+
+    virtual ISmartSocket *connect_timeout(unsigned timeoutms) override
+    {
+        ISocket *sock = nullptr;
+        SmartSocketEndpoint *ss = nullptr;
+        SocketEndpoint ep;
+        ISecureSocket *ssock = nullptr;
+        try
+        {
+            sock = connect_sock(timeoutms, ss, ep);
+            if (!sock)
+                throw createSmartSocketException(0, "securesmartsocket failed to connect");
+
+            ssock = secureContext->createSecureSocket(sock);
+
+            int status = ssock->secure_connect();
+            if (status < 0)
+            {
+                // secure_connect may also DBGLOG() errors ...
+                ssock->Release();
+                secureContext->Release();
+                throw MakeStringException(-1, "SSL smartsocket connect fail");
+            }
+            sock = ssock;
+
+            return new CSmartSocket(sock, ep, this);
+        }
+        catch (IException *e)
+        {
+            StringBuffer s("CSecureSmartSocketFactory::connect ");
+            ep.getUrlStr(s);
+            EXCLOG(e,s.str());
+            if (ss)
+                ss->status=false;
+            if (sock)
+                sock->Release();
+            throw;
+        }
```

Comment:
See other comments, the exception handling is already in base connect_sock and 'sock' would be invalid (already Released) by time it got here anyway.